### PR TITLE
Use the Actions artifacts client to avoid needing the `actions: read` permission

### LINKED ISSUE #284: Actions fail after update actions/deploy-pages to v4 and actions/upload-pages-artifact to v3
I just updated actions/deploy-pages to v4 and actions/upload-pages-artifact to v3 in [my repo](https://github.com/AREA44/astro-research-theme) and got this error:
```yml
Run actions/deploy-pages@v4
Fetching artifact metadata for github-pages in run 7264600[9](https://github.com/AREA44/astro-research-theme/actions/runs/7264600911/job/19792767463#step:2:10)[11](https://github.com/AREA44/astro-research-theme/actions/runs/7264600911/job/19792767463#step:2:12)
Creating Pages deployment with payload:
{
	"artifact_id": 1[12](https://github.com/AREA44/astro-research-theme/actions/runs/7264600911/job/19792767463#step:2:13)45035[14](https://github.com/AREA44/astro-research-theme/actions/runs/7264600911/job/19792767463#step:2:15),
	"pages_build_version": "cf67bf05e7d24c255c38726ac891fe4b93fbf25b",
	"oidc_token": "***"
}
Error: Creating Pages deployment failed
Error: HttpError: No artifact_url provided
    at /home/runner/work/_actions/actions/deploy-pages/v4/node_modules/@octokit/request/dist-node/index.js:124:1
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at createPagesDeployment (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/api-client.js:68:1)
    at Deployment.create (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/deployment.js:78:1)
    at main (/home/runner/work/_actions/actions/deploy-pages/v4/src/index.js:30:1)
Error: Error: Failed to create deployment (status: 400) with build version cf67bf05e7d24c255c38726ac891fe4b93fbf25b. Responded with: No artifact_url provided
```
Full log [here](https://github.com/AREA44/astro-research-theme/actions/runs/7264600911).

### LINKED ISSUE #285: Failing to fetch artifact metadata since 4.0.0
After updating to deploy-pages@v4 and upload-pages-artifact@v3, I am running into the following error:

```
Fetching artifact metadata for github-pages in run 7265993097
Error: Fetching artifact metadata failed. Is githubstatus.com reporting issues with API requests, Pages or Actions? Please re-run the deployment at a later time.
Error: HttpError: Resource not accessible by integration
    at /home/runner/work/_actions/actions/deploy-pages/v4/node_modules/@octokit/request/dist-node/index.js:124:1
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at getArtifactMetadata (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/api-client.js:10:1)
    at Deployment.create (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/deployment.js:66:1)
    at main (/home/runner/work/_actions/actions/deploy-pages/v4/src/index.js:30:1)
Error: Error: Failed to create deployment (status: 403) with build version 5d6a87698b916de4c981c3373b41af57cfb2983f. Ensure GITHUB_TOKEN has permission "pages: write".
```

> Is githubstatus.com reporting issues with API requests, Pages or Actions?

No.

> Please re-run the deployment at a later time.

I've tried running this multiple times throughout the day.

> Error: Failed to create deployment (status: 403) with build version 5d6a87698b916de4c981c3373b41af57cfb2983f. Ensure GITHUB_TOKEN has permission "pages: write".

Interestingly, I did not change anything about the token permissions. In fact, the `pages: write` permission is explicitly granted in the workflow:

```yaml
name: Deploy
on:
  push:
    branches:
      - "master"


jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - uses: actions/setup-node@v4
        with:
          node-version: "18.x"
          cache: "yarn"

      - run: yarn install

      - run: yarn generate

      - name: Upload artifact
        uses: actions/upload-pages-artifact@v3
        with:
          path: ./dist

  deploy:
    needs: build
    runs-on: ubuntu-latest
    permissions:
      pages: write
      id-token: write
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    steps:
      - name: Deploy
        id: deployment
        uses: actions/deploy-pages@v4
```

(This might be related to #284.)

### LINKED ISSUE #286: Bump to V4 broke the deploy step
Initial failure with permissions:
```
permissions:
  contents: read
  pages: write
  id-token: write
```
resulted in:
```
Fetching artifact metadata for github-pages in run 0000000000
Error: Fetching artifact metadata failed. Is githubstatus.com reporting issues with API requests, Pages or Actions? Please re-run the deployment at a later time.
Error: HttpError: Resource not accessible by integration
    at /home/runner/work/_actions/actions/deploy-pages/v4/node_modules/@octokit/request/dist-node/index.js:124:1
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at getArtifactMetadata (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/api-client.js:[10](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:11):1)
    at Deployment.create (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/deployment.js:66:1)
    at main (/home/runner/work/_actions/actions/deploy-pages/v4/src/index.js:30:1)
Error: Error: Failed to create deployment (status: 403) with build version 0000000000[15](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:16). Ensure GITHUB_TOKEN has permission "pages: write".
```

After updated the permissions to:
```
permissions:
  contents: read
  pages: write
  id-token: write
  actions: read
```

New failure:
```
Fetching artifact metadata for github-pages in run 0000000000
Creating Pages deployment with payload:
{
	"artifact_id": [11](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:12)0000000000,
	"pages_build_version": "0000000000",
	"oidc_token": 0000000000"
}
Error: Creating Pages deployment failed
Error: HttpError: No artifact_url provided
    at /home/runner/work/_actions/actions/deploy-pages/v4/node_modules/@octokit/request/dist-node/index.js:[12](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:13)4:1
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at createPagesDeployment (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/api-client.js:68:1)
    at Deployment.create (/home/runner/work/_actions/actions/deploy-pages/v4/src/internal/deployment.js:78:1)
    at main (/home/runner/work/_actions/actions/deploy-pages/v4/src/index.js:30:1)
Error: Error: Failed to create deployment (status: 400) with build version 0000000000[16](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:17)0000000000[19](https://github.com/0000000000/actions/runs/0000000000/job/0000000000#step:2:20)0000000000. Responded with: No artifact_url provided
```

### LINKED ISSUE #289: Remove requirement for `actions: read` permission
The release of `v4.0.0` added the use of a new public REST API that now requires the `actions: read` permission on the `GITHUB_TOKEN`.

However, if we use the `@actions/artifact` non package instead, that can utilize an internal API that will _not_ require that permission.

🔮 If we decide to allow Pages deployments from workflow runs separate from the artifact upload in the future, this `@actions/artifact` client will still work but then _will_ require the `actions: read` permission when used for such scenarios.

### PR DESCRIPTION
**Pull request by JamesMGreene**
- Fixes #284 
- Fixes #285 
- Fixes #286 
- Closes #287 
- Fixes #289

### CODE REVIEW

**Comment by ntkme** [comment]
@JamesMGreene My own test shows that `actions: read` does not need to be specified at all, and it was not the issue. The real issue is that `createPagesDeployment` is calling `POST /repos/{owner}/{repo}/pages/deployments` with artifact_id, but the API only takes artifact_url. Using actions artifacts client does not really change that behavior.

**Comment by kphrx** [comment]
`artifact_id` is a new api parameter used from #251. if reverting, maybe unnecessary v4.

**Comment by JamesMGreene** [comment]
Failures from a workflow run using `actions/deploy-pages@v4` without also granting `actions: read` permission:

permissions failure

After switching to use this PR branch for `actions/deploy-pages`, the same workflow succeeded without any errors. 🎉 🚀

File: src/internal/api-client.js [hunk 1]
```diff
@@ -1,50 +1,106 @@
...
+  const artifactClient = new DefaultArtifactClient()
+
+  // Primarily for debugging purposes, accuracy is not critical
+  const requestOptions = {
+    method: 'POST',
+    url: `${getArtifactsServiceOrigin()}/twirp/github.actions.results.api.v1.ArtifactService/ListArtifacts`,
+    headers: {
+      'content-type': 'application/json'
+    }
+  }
```
Comment (yoannchaudet) [inline src/internal/api-client.js:63]: Wow did not expect to see Twirp here! Is the client doing the auth for us or is it public 🤔 ?

File: src/internal/api-client.js [hunk 2]
```diff
@@ -1,50 +1,106 @@
...
+    const filteredArtifacts = response.data.artifacts.filter(artifact => artifact.name === artifactName)
 
-    const artifactCount = response.data.total_count
+    const artifactCount = filteredArtifacts.length
     core.debug(`List artifact count: ${artifactCount}`)
 
     if (artifactCount === 0) {
       throw new Error(
-        `No artifacts found for workflow run ${runId}. Ensure artifacts are uploaded with actions/artifact@v4 or later.`
+        'No artifacts found for this workflow run. Ensure artifacts are uploaded with actions/artifact@v4 or later.'
```
Comment (yoannchaudet) [inline src/internal/api-client.js:85]: nit: since we do filtering by name, we could adjust the wording a little

File: src/internal/api-client.js [hunk 3]
```diff
@@ -1,50 +1,106 @@
...
       )
     } else if (artifactCount > 1) {
       throw new Error(
-        `Multiple artifact unexpectedly found for workflow run ${runId}. Artifact count is ${artifactCount}.`
+        `Multiple artifacts unexpectedly found for this workflow run. Artifact count is ${artifactCount}.`
```
Comment (yoannchaudet) [inline src/internal/api-client.js:89]: Can two artifacts have the same name? (not that we should remove the code)

File: src/internal/api-client.js (hunk 1 above)
Comment (JamesMGreene) [reply src/internal/api-client.js:63]: The Artifacts/Results service Twirp calls are all handled in the `@actions/artifact` client (e.g. [`artifact-twirp-client.ts`](https://github.com/actions/toolkit/blob/main/packages/artifact/src/internal/shared/artifact-twirp-client.ts)

It works pretty similarly to when we were calling the Runtime service for artifact exchange URLs prior to `v4.0.0`.

**Review by yoannchaudet — DISMISSED** [review]
No blocking comments from me!

File: src/internal/api-client.js (hunk 1 above)
Comment (JamesMGreene) [reply src/internal/api-client.js:63]: P.S. This object is used strictly for debugging, in our case. It's effectively duplicating the internals of `@actions/artifact`, which I don't like buuuuuuut I want something there for troubleshooting. 🤷🏻

Comment (yoannchaudet) [reply src/internal/api-client.js:63]: Oki the take the token, got my question answered :)

File: src/internal/api-client.js (hunk 3 above)
Comment (JamesMGreene) [reply src/internal/api-client.js:89]: Yes, unfortunately, they can.

Alternatively, we could pass in a `latest: true` property and the `@actions/artifact` client to filter down to the most recently created artifact.

File: src/internal/api-client.js (hunk 2 above)
Comment (JamesMGreene) [reply src/internal/api-client.js:85]: Maybe something like:

```suggestion
        `No artifacts found for this workflow run with name "${artifactName}". Ensure artifacts are uploaded with actions/artifact@v4 or later.`
```

Comment (yoannchaudet) [reply src/internal/api-client.js]: ❤️ !

**Review by yoannchaudet — APPROVED** [review]

**Pull request merged**